products/bluebell/src/lexer.rs

```diff
@@ -0,0 +1,573 @@
+    chars: Peekable<CharIndices<'input>>,
+    /// A reference to the source file being tokenized.
+    document: &'input str,
+    /// The current line number being tokenized.
+    line: usize,
+    /// The current character number within the current line being tokenized.
+    character: usize,
+}
+
+impl<'input> Lexer<'input> {
+    pub fn new(input: &'input str) -> Self {
+        Lexer {
+            chars: input.char_indices().peekable(),
+            document: input,
+            line: 1,
+            character: 1,
+        }
+    }
+}
+
+impl<'input> Iterator for Lexer<'input> {
+    type Item = Spanned<Token<&'input str>, usize, ParseError>;
+
+    // <(usize, Token, usize, usize, usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
+
+
+        while let Some((start, ch)) = self.chars.next() {
+
+            let (token, end): (Token<&'input str>, usize) = {
+                let look_ahead = self.chars.peek().map(|(_, next_ch)| *next_ch);
+                self.character += ch.len_utf8();
+
+                let next_is_alpha_num_under = match look_ahead.map(|c| c.is_alphanumeric() || c == '_') {
+                        Some(true) => true,
+                        _ => false
+                    };
+
+                let next_is_numeric = match look_ahead.map(|c| c.is_numeric()) {
+                        Some(true) => true,
+                        _ => false
+                    };
+
+                // Handle more complex tokens, whitespace, and comments
+                if ch.is_whitespace() {
+                    if ch == '\n' {
+                        self.character = 0;
+                        self.line += 1;
+                    }
+                    continue;
+                } else if ch == '=' && look_ahead == Some('>') {
+                    self.chars.next();
+                    (Token::DoubleArrow, start + 2 * ch.len_utf8())
+                } else if ch == '-' && look_ahead == Some('>') {
+                    self.chars.next();
+                    (Token::Arrow, start + 2 * ch.len_utf8())
+                } else if ch == '-' && !next_is_numeric {
+                    (Token::Minus, start + ch.len_utf8())               
+                } else if ch == '<' && look_ahead == Some('-') {
+                    self.chars.next();
+                    (Token::LeftArrow, start + 2 * ch.len_utf8())
+                } else if ch == ':' && look_ahead == Some('=') {
+                    self.chars.next();
+                    (Token::ColonEquals, start + 2 * ch.len_utf8()) 
+                } else if ch == '_' && !next_is_alpha_num_under {
+                    (Token::Underscore, start + ch.len_utf8())
+                } else if ch == '(' && look_ahead == Some('*') {
+                    // Consume comment
+
+                    self.chars.next(); // Consume '*'
+                    let mut comment = String::new();
+
+                    while let Some((_, ch)) = self.chars.next() {
```

Comment:
Does this handle nested comments? AFAICT no?

i.e.
```
(* a comment (* an inner comment *)
```
should not parse, because the brackets are not balanced.

Reply:
This is not handled yet. As you will see from the comment following, the entire comment handling should be refactored since we want the comments in the AST such that we can reconstruct the code with comments from it (to make a formatter).

I'll postpone this for later.